### PR TITLE
Fix percentiles support in sqlite, mssql and mysql

### PR DESCRIPTION
**Pull request by serena-ruan**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Previously I was using the approximation method for these three types of engines for easier processing, however the assumption of linear distribution of data is not correct so we should calculate the real number of the percentile.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/db/utils.py [hunk 1]
```diff
@@ -401,4 +402,36 @@ def regexp(pattern, string):
 
             dbapi_conn.create_function("regexp", 2, regexp)
 
+            # Register PERCENTILE aggregate function (PERCENTILE_CONT with linear interpolation)
+            class PercentileAggregate:
```
Comment (TomeHirata) [inline mlflow/store/db/utils.py:406]: Can't we use the built-in `percentile` function? https://sqlite.org/percentile.html

Comment (serena-ruan) [reply mlflow/store/db/utils.py:406]: As for the availability:
```
The percentile extension has been part of [the amalgamation](https://sqlite.org/amalgamation.html) since SQLite version 3.51.0 (2025-11-04). However, the percentile extension is disabled by default and must be activated at compile-time using the [-DSQLITE_ENABLE_PERCENTILE](https://sqlite.org/compile.html#enable_percentile) compile-time option.
```
It requires special setup and doesn't work with older sqlite versions (2025-11-04 is pretty new), so this function registry setup is simpler and works for all versions

File: mlflow/store/db/utils.py [hunk 2]
```diff
@@ -401,4 +402,36 @@ def regexp(pattern, string):
...
+
+                def finalize(self):
+                    if not self.values or self.percentile_value is None:
+                        return None
+
+                    sorted_values = sorted(self.values)
+                    n = len(sorted_values)
+                    percentile_ratio = self.percentile_value / 100.0
+
+                    # Linear interpolation: index = percentile_ratio * (n - 1)
```
Comment (TomeHirata) [inline mlflow/store/db/utils.py:425]: Nice!

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 3]
```diff
@@ -113,33 +113,52 @@ class TraceMetricsConfig:
...
+                .within_group(column)
+                .over(partition_by=partition_by)
+            )
+        case "sqlite":
+            # SQLite percentile extension function (expects percentile as 0-100)
+            # Note: Requires the percentile extension to be loaded
+            # See: https://sqlite.org/percentile.html
+            return func.percentile(column, percentile_value)
+        case "mysql":
+            # MySQL 8.0+ supports PERCENT_RANK() function.
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py:165]: What happens if users are using MySQL 5.7 and tries to view trace dashboard?

File: mlflow/store/db/utils.py [hunk 4]
```diff
@@ -385,11 +385,12 @@ def create_sqlalchemy_engine(db_uri):
 
     engine = sqlalchemy.create_engine(db_uri, pool_pre_ping=True, **kwargs)
 
-    # Register REGEXP function for SQLite to enable RLIKE operator support
+    # Register custom functions for SQLite
     if db_uri.startswith("sqlite"):
 
         @event.listens_for(engine, "connect")
-        def _set_sqlite_regexp(dbapi_conn, connection_record):
+        def _set_sqlite_functions(dbapi_conn, connection_record):
```
[the remark this answers is not in the document]
Comment (TomeHirata) [reply mlflow/store/db/utils.py]: I think this suggestion makes sense

File: mlflow/store/db/utils.py (hunk 1 above)
Comment (TomeHirata) [reply mlflow/store/db/utils.py:406]: Make sense!

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 5]
```diff
@@ -113,33 +113,52 @@ class TraceMetricsConfig:
...
+    db_type: str, percentile_value: float, column, partition_by_columns: list[Column] | None = None
+):
     """
     Get percentile aggregation function based on database type.
 
     Args:
         db_type: Database type (e.g., "postgresql", "mssql", "mysql", "sqlite")
         percentile_value: Percentile value between 0 and 100 (e.g., 50 for median)
         column: SQLAlchemy column to compute percentile on
+        partition_by_columns: For MSSQL, columns to partition by in the OVER clause.
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py]: Can we document how this is used for MySQL?

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 6]
```diff
@@ -113,33 +113,52 @@ class TraceMetricsConfig:
...
-
-    Note:
-        - PostgreSQL and MSSQL use native PERCENTILE_CONT functions
-        - MySQL/SQLite uses an approximation with MAX/MIN weighted calculation
     """
-    percentile_value = percentile_value / 100
+    percentile_fraction = percentile_value / 100  # Convert to 0-1 range
+
     match db_type:
         case "postgresql":
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py]: btw, shall we define an enum for db types?

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 7]
```diff
@@ -186,14 +205,20 @@ def get_time_bucket_expression(
...
+    partition_by_columns: list[Column] | None = None,
+) -> Column:
     """
     Get the SQL aggregation expression for the given aggregation type and column.
 
     Args:
         aggregation: The aggregation of the metric
         db_type: Database type (for percentile calculations)
         column: The column to aggregate
+        partition_by_columns: For MSSQL percentile, columns to partition by in OVER clause
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py]: ditto

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 8]
```diff
@@ -490,28 +627,30 @@ def query_metrics(
...
+            expr = _get_aggregation_expression(agg, db_type, agg_column)
+            select_columns.append(expr.label(str(agg)))
+
+        query = query.with_entities(*select_columns)
+
+        # Extract underlying column expressions from labeled columns for GROUP BY/ORDER BY
+        if dimension_columns:
+            group_by_columns = [col.element for col in dimension_columns]
+            query = query.group_by(*group_by_columns)
+            # order by time bucket first, then by other dimensions
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py:660]: Where is it guaranteed that the time bucket is the first element of `dimension_columns` in the caller?

File: tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py [hunk 9]
```diff
@@ -2277,14 +2236,14 @@ def test_query_span_metrics_latency_multiple_aggregations(store: SqlAlchemyStore
...
-    expected_p95 = _get_expected_percentile_value(store, 95.0, 100.0, 500.0, latency_values)
+    expected_p50 = pytest.approx(np.percentile(latency_values, 50.0), abs=0.01)
+    expected_p95 = pytest.approx(np.percentile(latency_values, 95.0), abs=0.01)
 
     assert len(result) == 1
     assert asdict(result[0]) == {
         "metric_name": SpanMetricKey.LATENCY,
         "dimensions": {},
-        "values": {"AVG": 300.0, "P50": expected_p50, "P95": expected_p95},
+        "values": {"AVG": pytest.approx(300.0, abs=0.01), "P50": expected_p50, "P95": expected_p95},
```
Comment (TomeHirata) [inline tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py:2292]: Does AVG need `pytest.approx`?

File: tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py [hunk 10]
```diff
@@ -394,13 +362,17 @@ def test_query_trace_metrics_latency_multiple_aggregations(store: SqlAlchemyStor
...
-    expected_p99 = _get_expected_percentile_value(store, 99.0, 100.0, 500.0, values)
+    expected_p95 = pytest.approx(np.percentile(values, 95.0), abs=0.01)
+    expected_p99 = pytest.approx(np.percentile(values, 99.0), abs=0.01)
 
     assert asdict(result[0]) == {
         "metric_name": TraceMetricKey.LATENCY,
         "dimensions": {TraceMetricDimensionKey.TRACE_NAME: "workflow_a"},
-        "values": {"AVG": 300.0, "P95": expected_p95, "P99": expected_p99},
+        "values": {
+            "AVG": pytest.approx(300.0, abs=0.01),
```
[the remark this answers is not in the document]
Comment (TomeHirata) [reply tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py:418]: @serena-ruan can you take a look at this comment?

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 11]
```diff
@@ -113,33 +113,52 @@ class TraceMetricsConfig:
...
 
     Returns:
         SQLAlchemy aggregation function for percentile
-
-    Note:
-        - PostgreSQL and MSSQL use native PERCENTILE_CONT functions
-        - MySQL/SQLite uses an approximation with MAX/MIN weighted calculation
     """
-    percentile_value = percentile_value / 100
+    percentile_fraction = percentile_value / 100  # Convert to 0-1 range
```
Comment (TomeHirata) [inline mlflow/store/tracking/utils/sql_trace_metrics_utils.py:142]: Also, can we add an explanation on what interporation method is used in MLflow? I guess it's the same as linear interporation in the numpy doc? https://numpy.org/doc/stable/reference/generated/numpy.quantile.html

File: tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py [hunk 12]
```diff
@@ -338,11 +304,13 @@ def test_query_trace_metrics_latency_percentiles(
     )
 
     # Calculate expected values based on database type
-    expected_workflow_a = _get_expected_percentile_value(
-        store, percentile_value, 100.0, 300.0, [100.0, 200.0, 300.0]
+    expected_workflow_a = pytest.approx(
```
Comment (TomeHirata) [inline tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py:307]: Maybe worth adding a test case where the higher point and lower point are the same (e.g., median of [50,50]) as that's typically the edge case?

**Review by TomeHirata — APPROVED** [review]
Left some suggestions, but the logic overall LGTM!

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/store/tracking/utils/sql_trace_metrics_utils.py:165]: MySQL 8.0 is pretty old (5.7 already EOL in 2023)

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py [hunk 13]
```diff
@@ -447,6 +474,116 @@ def _apply_filters(query: Query, filters: list[str], view_type: MetricViewType)
...
+                    percentile_labels[str(agg)] = label
+        case "mysql":
+            # add single PERCENT_RANK column for interpolation
+            inner_columns.append(
+                func.percent_rank()
+                .over(partition_by=partition_by_columns, order_by=agg_column)
+                .label("_pct_rank")
+            )
+        case _:
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/store/tracking/utils/sql_trace_metrics_utils.py:544]: This is on-purpose since invalid_parameter_value indicates a user error, but this is not. In reality, this part is not reachable

File: mlflow/store/tracking/utils/sql_trace_metrics_utils.py (hunk 8 above)
Comment (serena-ruan) [reply mlflow/store/tracking/utils/sql_trace_metrics_utils.py:660]: Line 624 appends time bucket column to dimension_columns first

File: tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (serena-ruan) [reply tests/store/tracking/test_sqlalchemy_store_query_trace_metrics.py:418]: This is on-purpose since apparently AVG results of some databases have weird rounding behavior that may become 299.999999... In the dashboard we do rounding so there's no difference. For example:
```
>       assert asdict(result[0]) == {
            "metric_name": AssessmentMetricKey.ASSESSMENT_VALUE,
            "dimensions": {AssessmentMetricDimensionKey.ASSESSMENT_NAME: "accuracy"},
            "values": {"AVG": 0.85},
        }
E       AssertionError: assert {'dimensions'...000000000001}} == {'dimensions'...{'AVG': 0.85}}
E         
E         Omitting 2 identical items, use -vv to show
E         Differing items:
E         {'values': {'AVG': 0.8500000000000001}} != {'values': {'AVG': 0.85}}
E         Use -v to get more diff
```
For postgresql
